File: challenges/week4.cinema-seat-allocation/ikasty.scala
// Runtime: 1164 ms, faster than 100.00% of Scala online submissions
// Memory Usage: 64 MB, less than 100.00% of Scala online submissions
object Solution {
    def maxNumberOfFamilies(n: Int, reservedSeats: Array[Array[Int]]): Int = {
        val target = reservedSeats.map(_(0)).distinct
        val emptyline = (n - target.size) * 2
        val reserved = reservedSeats
            .filter(2 to 9 contains _(1))
            .groupBy(_(0))
        
        // seat 2, 3 => 0
        // seat 4, 5 => 1
        // seat 6, 7 => 2
        // seat 8, 9 => 3
        val results = for (row <- target) yield reserved
            .getOrElse(row, Array())
            .map(_(1) / 2 - 1).sorted.distinct match {
                case Array(0, 1)        => 1
                case Array(2, 3)        => 1
                case Array(0, 3)        => 1
                case x if x.size >= 2   => 0
                case x if x.size == 1   => 1
                case _                  => 2
            }

        emptyline + results.sum
    }
}

File: challenges/week4.cinema-seat-allocation/ikasty.py
# Runtime: 612 ms, faster than 93.80% of Python3 online submissions
# Memory Usage: 17.8 MB, less than 78.29% of Python3 online submissions
from collections import defaultdict

class Solution:
    def maxNumberOfFamilies(self, n: int, reservedSeats: List[List[int]]) -> int:
        reserved = defaultdict(int)
        for row, col in reservedSeats:
            if col in range(2, 10):
                reserved[row] |= 1 << (10 - col)

        group = (n - len(reserved)) * 2
        for row in reserved:
            if reserved[row] & 0b0000011110 == 0 or \
               reserved[row] & 0b0001111000 == 0 or \
               reserved[row] & 0b0111100000 == 0:
                group += 1
        return group
